frozen-abi/macro/src/lib.rs

```diff
@@ -28,41 +28,55 @@ pub fn derive_abi_enum_visitor(_item: TokenStream) -> TokenStream {
 }
 
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
-use proc_macro2::{Span, TokenStream as TokenStream2, TokenTree::Group};
+use proc_macro2::{Span, TokenStream as TokenStream2, TokenTree};
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
-use quote::quote;
+use quote::{quote, ToTokens};
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
 use syn::{
-    parse_macro_input, Attribute, AttributeArgs, Error, Fields, Ident, Item, ItemEnum, ItemStruct,
-    ItemType, Lit, Meta, NestedMeta, Variant,
+    parse_macro_input, Attribute, Error, Fields, Ident, Item, ItemEnum, ItemStruct, ItemType,
+    LitStr, Variant,
 };
 
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
 fn filter_serde_attrs(attrs: &[Attribute]) -> bool {
-    let mut skip = false;
-
-    for attr in attrs {
-        let ss = &attr.path.segments.first().unwrap().ident.to_string();
-        if ss.starts_with("serde") {
-            for token in attr.tokens.clone() {
-                if let Group(token) = token {
-                    for ident in token.stream() {
-                        if ident.to_string() == "skip" {
-                            skip = true;
-                        }
+    fn contains_skip(tokens: TokenStream2) -> bool {
+        for token in tokens.into_iter() {
+            match token {
+                TokenTree::Group(group) => {
+                    if contains_skip(group.stream()) {
+                        return true;
                     }
                 }
+                TokenTree::Ident(ident) => {
+                    if ident == "skip" {
+                        return true;
+                    }
+                }
+                TokenTree::Punct(_) => (),
+                TokenTree::Literal(_) => (),
             }
         }
+
+        false
     }
 
-    skip
+    for attr in attrs {
+        if !attr.path().is_ident("serde") {
+            continue;
+        }
+
+        if contains_skip(attr.to_token_stream()) {
+            return true;
+        }
+    }
```

Comment:
seems this loop equivalent to this simpler form?:

```rust
for attr in attrs {
    if attr.path().is_ident("serde") && contains_skip(attr.to_token_stream()) {
        return true;
    }
}
```

even more, you can just use `.any()`, right?:

```rust
attrs.iter().any(|attr| attr.path().is_ident("serde") && contains_skip(attr.to_token_stream()))
```

Reply:
Yes, but it makes it harder to read, I think :)
The first `if` says "only when we are looking at a serde attribute...".
And the second can be then understood in the context of the first.

`any()` works yes, but it pulls in so much syntactic noise: `.iter().any(|attr|` - I find it harder to understand, compared to a simple `for`/`if`.
I think, `for` and `if` statements should be used when they are a good fit.
Putting effort into avoiding them means requiring more effort from the reader when they are reading the code.
It is unclear to me what is the value in that.

I would change it, if you insist.
But I do not see any of the two proposed variants as improving readability.
In my mind, "dumb" code is easier to read :)

Reply:
hehe, seems you have a opinion. :) and i love these chats.

Firstly, I won't insist, after reading you thoughts. (after all, rust is multi-paradigm and why not foster diversity?)

> It is unclear to me what is the value in that.
> ...
> But I do not see any of the two proposed variants as improving readability.  In my mind, "dumb" code is easier to read :)

Actually, i  see a value. but, admittedly, it's subjective. at end of day, i think engineer's background could play a role to gauge the readability of code.

For me, I'm strongly influenced by dynamic scripting languages/functional languages. so, these iter heavy code looks more readable. that said, i think there's equally-justifiable audience who prefer simpler `for`/`if` (guessing pure C (or Go in recent times) strong background).

that said, i value shorter code for readability in terms of count of tokens and fewer control flow. so, I suggested the `.iter()`, combined with the aforementioned `.iter()` preference. (`.iter().any()` could even eat the `false` at the end of this fn.

Reply:
> hehe, seems you have a opinion. :) and i love these chats.

I'm relatively new to the Solana code base.  So I really appreciate your feedback.
I can certainly adjust to a different style, if there is a certain consensus.
But if not, I do have an opinion, yes %)

> Firstly, I won't insist, after reading you thoughts. (after all, rust is multi-paradigm and why not foster diversity?)
> 
> > It is unclear to me what is the value in that.
> > ...
> > But I do not see any of the two proposed variants as improving readability.  In my mind, "dumb" code is easier to read :)
> 
> Actually, i see a value. but, admittedly, it's subjective. at end of day, i think engineer's background could play a role to gauge the readability of code.

Hope they way I expressed my opinion was not too harsh.
Obviously, what you are saying has value.
I was trying to say that in my mind, there is no improvement in code readability.
This is the value I was talking about.
I think, writing it this way can be seen as a "smarter way".
It is interesting that you find a more "syntax heavy" code easier to read.

> For me, I'm strongly influenced by dynamic scripting languages/functional languages. so, these iter heavy code looks more readable. that said, i think there's equally-justifiable audience who prefer simpler `for`/`if` (guessing pure C (or Go in recent times) strong background).
> 
> that said, i value shorter code for readability in terms of count of tokens and fewer control flow. so, I suggested the `.iter()`, combined with the aforementioned `.iter()` preference. (`.iter().any()` could even eat the `false` at the end of this fn.

We are certainly bike shedding here :)
But with how distributed the team is, this discussion helps me understand what others working on this code think.

I wrote code in a number of different languages, including C++ and Haskell, which Rust seems to be mostly based on.
Each language seems to have a certain style that comes with it.
For Rust, I read the standard library for "style inspiration".

Iterators in Rust are trying to copy a lot of the Haskell list API ergonomics, but do it in an efficient C++ iterator style.
Yet, in my mind, if a `for`/`if`/`while` works nicely for a given case, it is a better choice.

In Haskell, it would be something like
```haskell
    any [ attr <- attrs
        , ident (path attr) == "serde"
        , containsSkip attr
        ]
```

Which is relatively low "noise".  And as there are no `for` loops in Haskell, an alternative would actually be pretty verbose.

But I do not think that

```rust
    attrs
        .iter()
        .any(|attr|
            attr.path().is_ident("serde")
            && contains_skip(attr.to_token_stream())
        )
```
Is comparable in readability.